grafana-watcher: retry failed updates from ConfigMap changes

samuelmanzer opened the pull request:
Currently, an update will only be partially applied if an
error occurs in one of the handler routines. The grafana
system will be left in a dirty intermediate state until
the ConfigMap is updated again. This change will retry failed updates
until all scheduled updates have succeeded OR until the ConfigMap
is changed again.

This log file from the `grafana-watcher` container in the `grafanaXXX` pod from a `kube-prometheus` deployment shows the retry loop working correctly (I have no idea why I get so many timouts T_T):

```
2017/08/04 20:45:25 Waiting for Grafana to be available.
2017/08/04 20:45:41 Initializing datasources.
2017/08/04 20:45:41 Retrieving existing datasources
2017/08/04 20:45:41 Deleting 0 datasources
2017/08/04 20:45:41 Creating datasource from: /var/grafana-dashboards/prometheus-datasource.json
2017/08/04 20:45:42 Initializing dashboards.
2017/08/04 20:45:42 Retrieving existing dashboards
2017/08/04 20:45:58 Deleting 0 dashboards
2017/08/04 20:45:58 Creating dashboard from: /var/grafana-dashboards/all-nodes-dashboard.json
2017/08/04 20:45:58 Creating dashboard from: /var/grafana-dashboards/deployment-dashboard.json
2017/08/04 20:46:24 Creating dashboard from: /var/grafana-dashboards/kubernetes-pods-dashboard.json
2017/08/04 20:46:50 Creating dashboard from: /var/grafana-dashboards/node-dashboard.json
2017/08/04 20:47:06 Creating dashboard from: /var/grafana-dashboards/resource-requests-dashboard.json
2017/08/04 20:47:12 Starting...
2017/08/04 20:48:32 ConfigMap modified
2017/08/04 20:48:32 Retrieving existing dashboards
2017/08/04 20:49:02 error: Get http://admin:admin@localhost:3000/api/search: dial tcp: i/o timeout
2017/08/04 20:49:02 Retrieving existing datasources
2017/08/04 20:49:17 Deleting 1 datasources
2017/08/04 20:49:17 Deleting datasource: 1
2017/08/04 20:49:18 Creating datasource from: /var/grafana-dashboards/prometheus-datasource.json
2017/08/04 20:49:29 Retrieving existing dashboards
2017/08/04 20:49:44 Deleting 5 dashboards
2017/08/04 20:49:44 Deleting dashboard: all-nodes
2017/08/04 20:49:45 Deleting dashboard: deployment
2017/08/04 20:50:10 error during retry attempt: Delete http://admin:admin@localhost:3000/api/dashboards/db/deployment: dial tcp [::1]:3000: connect: cannot assign requested address
2017/08/04 20:50:20 Retrieving existing dashboards
2017/08/04 20:50:31 Deleting 4 dashboards
2017/08/04 20:50:31 Deleting dashboard: deployment
2017/08/04 20:50:31 Deleting dashboard: nodes
2017/08/04 20:51:01 error during retry attempt: Delete http://admin:admin@localhost:3000/api/dashboards/db/nodes: dial tcp: i/o timeout
2017/08/04 20:51:11 Retrieving existing dashboards
2017/08/04 20:51:41 error during retry attempt: Get http://admin:admin@localhost:3000/api/search: dial tcp: i/o timeout
2017/08/04 20:51:51 Retrieving existing dashboards
2017/08/04 20:52:21 error during retry attempt: Get http://admin:admin@localhost:3000/api/search: dial tcp: i/o timeout
2017/08/04 20:52:31 Retrieving existing dashboards
2017/08/04 20:52:57 Deleting 3 dashboards
2017/08/04 20:52:57 Deleting dashboard: nodes
2017/08/04 20:52:57 Deleting dashboard: pods
2017/08/04 20:53:18 Deleting dashboard: resource-requests
2017/08/04 20:53:48 error during retry attempt: Delete http://admin:admin@localhost:3000/api/dashboards/db/resource-requests: dial tcp: i/o timeout
2017/08/04 20:53:58 Retrieving existing dashboards
2017/08/04 20:54:24 Deleting 1 dashboards
2017/08/04 20:54:24 Deleting dashboard: resource-requests
2017/08/04 20:54:25 Creating dashboard from: /var/grafana-dashboards/all-nodes-dashboard.json
2017/08/04 20:54:55 Creating dashboard from: /var/grafana-dashboards/deployment-dashboard.json
2017/08/04 20:55:15 Creating dashboard from: /var/grafana-dashboards/kubernetes-pods-dashboard.json
2017/08/04 20:55:45 Creating dashboard from: /var/grafana-dashboards/node-dashboard.json
2017/08/04 20:56:06 Creating dashboard from: /var/grafana-dashboards/resource-requests-dashboard.json
```

brancz commented on contrib/grafana-watcher/main.go:
```diff
@@ -57,22 +57,50 @@ func (w *volumeWatcher) Run() {
+				}
+				handlersToRetry = remainingHandlers
+				// don't excessively spam retries
+				time.Sleep(5 * time.Second)
```
for this we should use a ticker

samuelmanzer replied:
I have one concern about using the ticker instead of `default`; `default` will always yield to new watcher updates, whereas the ticker may not yield for a few iterations because  the `select` chooses randomly among actionable cases (source: https://golang.org/ref/spec#Select_statements).  Let me know your thoughts, I have the ticker version ready to go as well so I'm ok with either.

brancz commented:
Due to some changes in our CI infrastructure this will also need a rebase.

samuelmanzer commented:
thanks for heads up - will rebase changes tonight

samuelmanzer commented:
Updated to use a ticker (though `default` has one advantage as mentioned in my other reply, let me know your thoughts on that). Confirmed still works by killing `grafana-server` immediately after ConfigMap update to force some operations to fail: 

```
2017/08/11 19:30:13 ConfigMap modified
2017/08/11 19:30:13 Retrieving existing dashboards
2017/08/11 19:30:13 error: Get http://admin:admin@localhost:3000/api/search: dial tcp [::1]:3000: getsockopt: connection refused
2017/08/11 19:30:13 Retrieving existing datasources
2017/08/11 19:30:13 error: Get http://admin:admin@localhost:3000/api/datasources: dial tcp [::1]:3000: getsockopt: connection refused
2017/08/11 19:30:20 Retrying 2 previously failed operations...
2017/08/11 19:30:20 Retrieving existing dashboards
2017/08/11 19:30:21 Deleting 0 dashboards
2017/08/11 19:30:21 Creating dashboard from: /var/grafana-dashboards/all-nodes-dashboard.json
2017/08/11 19:30:22 Creating dashboard from: /var/grafana-dashboards/deployment-dashboard.json
2017/08/11 19:30:22 Creating dashboard from: /var/grafana-dashboards/kubernetes-pods-dashboard.json
2017/08/11 19:30:23 Creating dashboard from: /var/grafana-dashboards/node-dashboard.json
2017/08/11 19:30:24 Creating dashboard from: /var/grafana-dashboards/resource-requests-dashboard.json
2017/08/11 19:30:24 Retrieving existing datasources
2017/08/11 19:30:25 Deleting 0 datasources
2017/08/11 19:30:25 Creating datasource from: /var/grafana-dashboards/prometheus-datasource.json
```

samuelmanzer commented:
ok, after latest rebase should be conflict-free and ready to go - let me know if anything else needed here

samuelmanzer commented:
@brancz does this look good now?

brancz commented:
Sorry for not getting back to this! The code lgtm. Could you just bump the grafana-watcher version used throughout this repo, so I can just publish the new image right before I merge this PR?

samuelmanzer commented:
Ok, should be bumped everywhere, looks like these were the 3 where bump needed: 

```
	modified:   contrib/grafana-watcher/examples/grafana-bundle.yaml
	modified:   contrib/kube-prometheus/manifests/grafana/grafana-deployment.yaml
	modified:   helm/grafana/templates/grafana-deployment.yaml
```

brancz commented:
Perfect! Will push the image and then merge the PR. Thanks a lot for all your efforts!
